File: leet-code/constructFromPrePost.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

function TreeNode (val) {
  this.val = val
  this.left = this.right = null
}
/*
 * @param {number[]} pre
 * @param {number[]} post
 * @return {TreeNode}
 */
var constructFromPrePost = function (pre, post) {
  let len = pre.length
  if (len === 0) {
    return null
  }
  let root = new TreeNode(pre[0])
  if (len === 1) {
    return root
  }

  let L = 0
  for (let i = 0; i < post.length; i++) {
    if (post[i] === pre[1]) {
      L = i + 1
    }
  }

  root.left = constructFromPrePost(pre.slice(1, L + 1), post.slice(0, L))
  root.right = constructFromPrePost(
    pre.slice(L + 1, len),
    post.slice(L, len - 1)
  )
  return root
}

console.log(constructFromPrePost([1, 2, 4, 5, 3, 6, 7], [4, 5, 2, 6, 7, 3, 1]))

File: leet-code/constructFromPrePost.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func constructFromPrePost(pre []int, post []int) *TreeNode {
	pLen := len(pre)
	if pLen == 0 {
		return nil
	}
	var root *TreeNode = new(TreeNode)
	root.Val = pre[0]
	if pLen == 1 {
		return root
	}

	N := -1
	for i := range post {
		if post[i] == pre[1] {
			N = i + 1
		}
	}

	root.Left = constructFromPrePost(pre[1:N+1], post[0:N])
	root.Right = constructFromPrePost(pre[N+1:pLen], post[N:len(post)-1])
	return root
}
